core/src/main/scala/kafka/server/DelayedOperation.scala

```diff
@@ -292,34 +309,46 @@ class DelayedOperationPurgatory[T <: DelayedOperation](purgatoryName: String,
    */
   private class Watchers(val key: Any) {
 
-    private[this] val operations = new LinkedList[T]()
+    private[this] val operations = new ConcurrentLinkedQueue[T]()
+    private[this] val approximateWatched = new AtomicInteger(0)
 
-    def watched: Int = operations synchronized operations.size
+    def watched: Int = approximateWatched.get
+
+    def isEmpty: Boolean = operations.isEmpty
 
     // add the element to watch
     def watch(t: T) {
-      operations synchronized operations.add(t)
+      operations.add(t)
+      approximateWatched.incrementAndGet()
+    }
+
+    private def purge(t: T): Boolean = {
+      if (t.purge()) {
+        approximateWatched.decrementAndGet()
+        true
+      } else {
+        false
+      }
     }
 
     // traverse the list and try to complete some watched elements
     def tryCompleteWatched(): Int = {
-
       var completed = 0
-      operations synchronized {
-        val iter = operations.iterator()
-        while (iter.hasNext) {
-          val curr = iter.next()
-          if (curr.isCompleted) {
-            // another thread has completed this operation, just remove it
-            iter.remove()
-          } else if (curr synchronized curr.tryComplete()) {
+      val iter = operations.iterator()
+      while (iter.hasNext) {
+        val curr = iter.next()
+        if (curr.isCompleted) {
+          // another thread has completed this operation, just remove it
+          iter.remove()
+          purge(curr)
+        } else if (curr.safeTryComplete()) {
+          iter.remove()
+          if (purge(curr))
             completed += 1
```

Comment:
I am not sure why we need the `purge` check. When we call `tryComplete` it should only return `true` if it was completed by the caller. If it was completed by another thread, it should return `false`. This is specified by `forceComplete` and the implementation seems fine:

```scala
def forceComplete(): Boolean = {
    if (completed.compareAndSet(false, true)) {
      // cancel the timeout timer
      cancel()
      onComplete()
      true
    } else {
      false
    }
```

Also, it's interesting that the return value of `tryCompleteWatched` and `purgeCompleted` is only used for `debug` logging, which is usually disabled anyway.